### ISSUE TITLE
New Post Processor System for New Impact Function

### ISSUE THREAD

**Issue by ismailsunni** [inasafe/inasafe#3208]
### Problem

Since we will move to the next version of impact function architecture, we need to update or create new post processor system. This ticket will be the parent ticket of all post processors tickets
### Proposed Solution
1. We will need a new keyword to store the required regarding to post processor. In the current architecture, we already have building_type and road_type field in the exposure layer to store the type of building and road. We will do the similar things. We also have women, youth, adult, and elderly field in the aggregation layer to store the ratio of each things. I suggest we store them in one keyword that contains a dictionary to make the metadata shorter and we can add more post processor in the future without changing the metadata scheme.
   Sample proposed implementation:
   
   ``` python
   inasafe_fields = {
   'type': 'TYPE',
   'population': 'pop_1',
   'value': 'value',
   'rate': 'rate',
   'women_ratio': 'women',
   'youth_ratio': 'youth',
   'adult_ratio': 'adult',
   'elderly_ratio': 'elderly'
   }
   ```
2. We will define those field details in the `definitions.py` From here, we can generate the name, description, field type (?), required or not, exposure or hazard type (?). Sample proposed implementation:
   
   ``` python
   population_field = {
   'key': 'population',
   'name': tr('Population'),
   'type': [float, int],
   'layer_purpose': 'exposure',
   'description': tr('Attribute for the number of population in the feature')
   }
   
   ```
   
   And we will put them in list of inasafe fields so we can iterate through them:
   
   ``` python
   inasafe_fields = [
   population_field,
   type_field,
   value_field,
   rate_field
   ]
   ```
3. For some fields, we can generate on the fly if we can't specify them in the original layer (e.g. population field in population raster, gender ratio in aggregation layer). We can use default value or use specific process to create and populate the field.
4. We will define all post processor in the definitions.py. 
   
   ``` python
   post_processor_gender = {
   'key': 'gender',
   'name': tr('Gender'),
   'description': tr(
       'Post processor to calculate the number of affected women'),
   'input': [
       population_field,
       gender_ratio_field
   ],
   'output': {
       'women': {
           'default': 'women',
           'description': tr('Number of affected women.'),
           'formula': 'population_field * gender_ratio_field'
       }
   }
   }
   ```
   
   We will put them in a list then iterate them, doing the post processing one by one, whenever possible. I am planning to put the calculation in the definitions.py, but I haven't got brilliant idea yet.
   
   ``` python
   post_processors = [
   post_processor_gender.
   post_processor_age
   ]
   ```
5. For the post processor result, we should also put some extra keywords so that we can recognize which field is what in the impact layer. I propose that we will do it like in the `inasafe_fields`. We can call it `post_processor_fields`.
   
   ``` python
   post_processor_fields = {
   'gender': 'women',
   'youth': 'youth',
   'adult': 'adult'
   }
   ```

Comments and suggestions are welcome
### Related Tickets
#3193
#3194
#3195
#3196
#3197
#3198
#3199
#3200

cc @timlinux @Gustry @lucernae @Charlotte-Morgan

**Comment by timlinux** [comment]
Hi @ismailsunni. I like where you are going with this. A few small comments:

* young_ratio -> youth_ratio
* in your second snippet, you have the layer keyword twice - can you clarify which it would be? And should we not use matching terminology from the layer (purpose: hazard, purpose: exposure, purpose: aggregation)?
* woman -> use the plural women in the above examples
* 'formula': 'population_field * gender_ratio_field' - how were you thinking of implementing this? a simple expression parser?
* What would we do when there is no ratio data in the aggregation layer? I think we can write it into the aggregate hazard layer using system defaults if it is not explicitly provided?

**Comment by Gustry** [comment]
I like it too.
If you start writing this, we should create a new file 'postprocessor' in @timlinux 's branch #3205 about refactoring definitions.py

**Comment by ismailsunni** [comment]
Hi @timlinux 
I have addressed the first 3 items of your comments, all are applied. Thanks
The 4th one, I have looked for that, but haven't look deeper. But yes, simple expression parser is my simplest solution.
The 5th one, yes, that's exactly what I will do. See point 3 in the post above.

Hi @Gustry Sure. I will write that while creating the post processor code.

Thanks guys for the comment. Been a while since there is a discussion in InaSAFE's thread :D
